bluepyentity/register.py

```diff
@@ -0,0 +1,472 @@
+ENTITIES_PATH = pkg_resources.resource_filename(__name__, "entity_definitions.yaml")
+
+
+def get_default_params(type_):
+    """Get default parameters for given type.
+
+    Args:
+        type_ (str): type of resource
+
+    Returns:
+        dict: default parameters as dict
+    """
+    return utils.parse_dict_from_file(DEFAULT_PARAMS_PATH).get(type_, {})
+
+
+def get_entity_definitions():
+    """Get all known entities"""
+    entities = utils.parse_dict_from_file(ENTITIES_PATH)
+    skipped_entities = [
+        "_Entity",
+        "EntityMixin",
+    ]
+    for se in skipped_entities:
+        if se in entities:
+            del entities[se]
+    return entities
+
+
+def get_entity_definition(type_):
+    """Get entity definition from the parsed definition file."""
+    entities = utils.parse_dict_from_file(ENTITIES_PATH)
+
+    def get_entity(name):
+        ent_dict = entities[name].copy()
+        ids = set(ent_dict.pop("_id_fields", []))
+
+        for ent in ent_dict.pop("_inherits", []):
+            inherited_ent, inherited_ids = get_entity(ent)
+            ids.update(inherited_ids)
+            ent_dict = dict(ent_dict, **inherited_ent)
+
+        return ent_dict, ids
+
+    if type_ in entities:
+        schema, ids = get_entity(type_)
+        return {"schema": schema, "ids": ids}
+
+    return None
+
+
+def _is_url(item):
+    return isinstance(item, str) and RE_URL.match(item)
+
+
+def _is_supported_type(type_):
+    # This is a bit dangerous as the type can be _Entity which would evaluate as True here
+    return get_entity_definition(type_) is not None
+
+
+def _wrap_linked_file_path(path):
+    """Wraps path as file URI if not already wrapped"""
+    return RE_NOT_FILE_URI.sub(FILE_URI, path)
+
+
+def _wrap_id(id_, type_):
+    """Wrap Nexus ID to an object"""
+    return {"type": type_, "id": id_}
+
+
+def _wrap_id_fetch(id_, forge):
+    """Find items with given IDs in Nexus and wrap them as objects."""
+    if _is_url(id_):
+        try:
+            return _wrap_id(id_, forge.retrieve(id_, cross_bucket=True).type)
+        except kgforge.core.commons.exceptions.RetrievalError as e:
+            raise RuntimeError(e) from e
+    elif isinstance(id_, list):
+        return [_wrap_id_fetch(i, forge) for i in id_]
+
+    return id_
+
+
+# NOTE: Should this work with bluepyentity.nexus.entityEntity?
+class Resource:
+    """Base class for resources to register."""
+
+    # This value should came from the schemas
+    required = set()
+
+    def __init__(self, definition, forge):
+        """Resource initializer.
+
+        Args:
+            definition (dict): attributes and their values to register
+            forge (kgforge.core.KnowledgeGraphForge): nexus-forge instance
+        """
+        # TODO: add registration to nexus.connector.Connector and use that instead of using
+        # kgforge methods directly
+        self._definition = definition
+        self._forge = forge
```

Comment:
This is a corner case, but relying on a forge instance for the entire lifetime of the instance may result in the token expiring while the object is still in use.

I can see two possible ways to address that:

Wrap forge methods to check for token expiration. Something similar to what I am doing here (adapting what Genrich is doing in entity-management for forge instantiation):
https://bbpgitlab.epfl.ch/nse/cwl-registry/-/blob/main/src/cwl_registry/nexus.py#L92
but for each method that talks to nexus, which is a challenging task given how convoluted forge is.

We could abstract away forge so that the direct API can also be used. Of course, for that to work, we would need to rely on to forge specifics as less as possible.

I am not sure if this is something we want to handle but in any case, I am leaving it here to have it in the back of our minds. And of course in order to be able to refresh tokens automatically we will also need to handle refresh/offline tokens.

Reply:
Yeah, I think the safest for now is to have functions to create and register instances and leave forge out of the class itself. 

i.e., `resource = create_resource(definition, forge_instance)` and `register(resource, forge_instance)`

that way the forge instance is easier to replace/renew session, etc. What do you think?